stan/math/rev/fun/divide.hpp

```diff
@@ -10,171 +10,41 @@
- * @return matrix divided by the scalar
- */
-template <typename Mat, typename = require_eigen_vt<std::is_arithmetic, Mat>>
-inline auto divide(const Mat& m, const var& c) {
-  auto* baseVari
-      = new internal::matrix_scalar_divide_dv_vari<Mat::RowsAtCompileTime,
-                                                   Mat::ColsAtCompileTime>(m,
-                                                                           c);
-  Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime> result(
-      m.rows(), m.cols());
-  result.vi()
-      = Eigen::Map<matrix_vi>(baseVari->adjResultRef_, m.rows(), m.cols());
-  return result;
-}
-
-/**
- * Return matrix divided by scalar.
- *
- * @tparam Mat type of the matrix or expression
- * @param[in] m specified matrix or expression
- * @param[in] c specified scalar
- * @return matrix divided by the scalar
- */
-template <typename Mat, typename = require_eigen_vt<is_var, Mat>>
-inline auto divide(const Mat& m, const double& c) {
-  auto* baseVari
-      = new internal::matrix_scalar_divide_vd_vari<Mat::RowsAtCompileTime,
-                                                   Mat::ColsAtCompileTime>(m,
-                                                                           c);
-  Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime> result(
-      m.rows(), m.cols());
-  result.vi()
-      = Eigen::Map<matrix_vi>(baseVari->adjResultRef_, m.rows(), m.cols());
-  return result;
-}
 
 /**
  * Return matrix divided by scalar.
  *
  * @tparam Mat type of the matrix or expression
+ * @tparam Scal type of the scalar
  * @param[in] m specified matrix or expression
  * @param[in] c specified scalar
  * @return matrix divided by the scalar
  */
-template <typename Mat, typename = require_eigen_vt<is_var, Mat>,
-          typename = void>
-inline auto divide(const Mat& m, const var& c) {
-  auto* baseVari
-      = new internal::matrix_scalar_divide_vv_vari<Mat::RowsAtCompileTime,
-                                                   Mat::ColsAtCompileTime>(m,
-                                                                           c);
-  Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime> result(
-      m.rows(), m.cols());
-  result.vi()
-      = Eigen::Map<matrix_vi>(baseVari->adjResultRef_, m.rows(), m.cols());
-  return result;
+template <typename Mat, typename Scal, require_eigen_t<Mat>* = nullptr,
+          require_stan_scalar_t<Scal>* = nullptr,
+          require_any_st_var<Mat, Scal>* = nullptr>
+inline Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime>
+divide(const Mat& m, Scal c) {
+  auto arena_m = to_arena_if<!is_constant<Mat>::value>(m);
+
+  double invc = 1.0 / value_of(c);
+
+  using Mat_v
+      = Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime>;
+
+  arena_matrix<Mat_v> res = invc * value_of(m);
+
+  reverse_pass_callback([=]() mutable {
+    if (!is_constant<Mat>::value) {
+      accumulate_adjoints(arena_m, invc * res.adj());
```

Comment:
arena_m might be an Eigen::MatrixXd or it might be an `arena_matrix<Eigen::Matrix<var, R, C>>` so I gotta have a function here that does different things for the different types.

The way you currently have dot_product is different, and that again is different from the way you originally had dot_product (with values and vars separate and explicitly typed so this wouldn't be necessary).

This code doesn't seem great to me but I dunno which way we should go. Thoughts?

Reply:
I am solving these kind of problems using `forward_as`. It allows you to specify which type you are expecting in a particular if with a type-dependent condition. Its return type will be what you specified, so the code can always compile. It should only be used in such ifs, so that it is actually called only with a compatible type.

```suggestion
      forward_as<Mat_v>(arena_m).adj() += invc * res.adj();
```